### PR TITLE
Glance Card Entity Options

### PR DESCRIPTION
**Pull request by zsarnett**
First PR to add this functionality. Will work later on design

[Screenshot from Gyazo](https://gyazo.com/9e2c44094f61bd380c53dd8c1a3cef08)

### CODE REVIEW

File: src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import { TemplateResult } from "lit-html";
+import "@polymer/paper-dropdown-menu/paper-dropdown-menu";
+import "@polymer/paper-item/paper-item";
+import "@polymer/paper-listbox/paper-listbox";
+
+import { HuiEntityEditor } from "../../../components/hui-entity-editor";
+import { HomeAssistant } from "../../../../../types";
+import { ConfigEntity } from "../../../cards/hui-glance-card";
+
+export class HuiGlanceEntitySelect extends HuiEntityEditor {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts]: Let's keep the names the same
```suggestion
export class HuiGlanceEntityEditor extends HuiEntityEditor {
```

File: src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  protected renderOptions(
+    entityConf: ConfigEntity,
+    index: number
+  ): TemplateResult {
+    return html`
+      <paper-toggle-button>Options</paper-toggle-button>
+      <div class="options">
+        <paper-input
+          label="Name"
+          .value="${entityConf.name || ""}"
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts]: Check how we did this at the editor: https://github.com/home-assistant/home-assistant-polymer/blob/dev/src/panels/lovelace/editor/config-elements/hui-glance-card-editor.ts#L60

File: src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+          .configValue="${"name"}"
+          .index="${index}"
+          @value-changed="${this._optionChanged}"
+        ></paper-input>
+        <paper-input
+          label="Icon"
+          .value="${entityConf.icon || ""}"
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts]: Same

File: src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+          .configValue="${"icon"}"
+          .index="${index}"
+          @value-changed="${this._optionChanged}"
+        ></paper-input>
+        <paper-dropdown-menu label="Toggle Action">
+          <paper-listbox
+            slot="dropdown-content"
+            .selected="${entityConf.tap_action || "more-info"}"
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts]: Same

File: src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+          >
+            <paper-item action="more-info">More Info Dialog</paper-item>
+            <paper-item action="toggle">Toggle</paper-item>
+            <paper-item action="call-service">Call Service</paper-item>
+          </paper-listbox>
+        </paper-dropdown-menu>
+        <paper-dropdown-menu label="Hold Action">
+          <paper-listbox
+            slot="dropdown-content"
+            .selected="${entityConf.hold_action || ""}"
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts]: Same

File: src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+          >
+            <paper-item action="">No Action</paper-item>
+            <paper-item action="more-info">More Info Dialog</paper-item>
+            <paper-item action="toggle">Toggle</paper-item>
+            <paper-item action="call-service">Call Service</paper-item>
+          </paper-listbox>
+        </paper-dropdown-menu>
+        <paper-input
+          label="Service"
+          .value="${entityConf.service || ""}"
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts]: Same

File: src/panels/lovelace/editor/config-elements/hui-glance-card-editor.ts [hunk 7]
```diff
@@ -14,9 +14,9 @@ import { fireEvent } from "../../../../common/dom/fire_event";
 import { Config, ConfigEntity } from "../../cards/hui-glance-card";
 import { configElementStyle } from "./config-elements-style";
 
+import "./entity-selectors/hui-glance-entity-select";
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-glance-card-editor.ts]: Let's make a subfolder for every card 
```suggestion
import "./glance-card/hui-glance-entity-select";
```

File: src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts (hunk 2 above)
Comment (bramkragten) [reply src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts]: And then filter it: https://github.com/home-assistant/home-assistant-polymer/blob/dev/src/panels/lovelace/editor/config-elements/hui-glance-card-editor.ts#L107

File: src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts (hunk 1 above)
Comment (balloob) [reply src/panels/lovelace/editor/config-elements/entity-selectors/hui-glance-entity-select.ts]: make sure the filename matches too

File: src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts [hunk 8]
```diff
@@ -87,17 +86,15 @@ export class HuiGlanceEntityEditor extends HuiEntityEditor {
...
-        target.value === this._glanceOptions!._name) ||
-      (target.configValue! === "icon" &&
-        target.value === this._glanceOptions!._icon) ||
-      (target.configValue! === "tap_action" &&
-        target.value === this._glanceOptions!._tapaction) ||
-      (target.configValue! === "hold_action" &&
-        target.value === this._glanceOptions!._holdaction) ||
-      (target.configValue! === "service" &&
-        target.value === this._glanceOptions!._service)
+      (target.configValue! === "name" && value === entity.name) ||
```
[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: Damn. I wrote it that way first and then redid it... FML

[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: Actually I wrote it that way before this last commit. Is that what you were thinking? 

Making it a property

[the remark this answers is not in the document]
Comment (bramkragten) [reply src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: Yeah, looks better, why did you change that? 😜

[the remark this answers is not in the document]
Comment (bramkragten) [reply src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: But that property would be overwritten for every entity, so you should combine both of them. Get the entity based in the index.

[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: Thats where I dont understand why we would use a Property then and compare against the entities Index. I dont see the point if we arent using it in optionChanged. Unless you are saying 

```
    const entity = this.entities![target.index!];
    this._glanceOptions = new GlanceOptions(entity);
```

[the remark this answers is not in the document]
Comment (bramkragten) [reply src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: There is no need for a property, you can instantiate the GlanceOptions class again in the changed handler, or the property should be an array of GlanceOptions.

File: src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+            .index="${index}"
+            @selected-item-changed="${this._optionChanged}"
+          >
+            <paper-item action="">No Action</paper-item>
+            <paper-item action="more-info">More Info Dialog</paper-item>
+            <paper-item action="toggle">Toggle</paper-item>
+            <paper-item action="call-service">Call Service</paper-item>
+          </paper-listbox>
+        </paper-dropdown-menu>
+        <paper-input
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: We should only show this when one of the actions is service?
```suggestion
        ${
          glanceOptions._holdaction === "call-service" ||
          glanceOptions._tapaction === "call-service"
            ? html`
                <paper-input
                  label="Service"
                  .value="${glanceOptions._service}"
                  .configValue="${"service"}"
                  .index="${index}"
                  @value-changed="${this._optionChanged}"
                ></paper-input>
              `
            : ""
        }
```

File: src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import { ConfigEntity } from "../../../cards/hui-glance-card";
+
+export class GlanceOptions {
```
Comment (balloob) [inline src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts]: If you need to do this for many values, you can also do this:

```ts
const values = {
  name: "",
  icon: "",
  ...entityConf
}
```

Comment (balloob) [reply src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts]: Or something like this:

```ts
const defaults = { name: "", icon: "" }
const values = {...defaults, ...entityConf }
```

Comment (balloob) [reply src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts]: (these are just ideas, the current approach is also fine)

File: src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    const glanceOptions = new GlanceOptions(entityConf);
+
+    return html`
+      <paper-toggle-button>Options</paper-toggle-button>
+      <div class="options">
+        <paper-input
+          label="Name"
+          .value="${glanceOptions._name}"
+          .configValue="${"name"}"
+          .index="${index}"
```
Comment (balloob) [inline src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: why not just set `.index` on the parent div? That way you set it once.

File: src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  private entityConf?: ConfigEntity;
+
+  constructor(entityConf: ConfigEntity) {
+    this.entityConf = entityConf;
+  }
+
+  get _name(): string {
```
Comment (balloob) [inline src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts]: Don't prefix it with `_`, that indicates it's for private use within the class, and that's not here

File: src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    `;
+  }
+
+  protected _optionChanged(ev: Event): void {
+    const target = ev.target! as EditorTarget;
+    const value = target.value || target.selected;
+    const entity = this.entities![target.index!];
+    const glanceOptions = new GlanceOptions(entity);
+
+    if (
```
Comment (balloob) [inline src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]:
```ts
if (glanceOptions[target.configValue] === value) { return; }
```

File: src/panels/lovelace/components/hui-entity-editor.ts [hunk 14]
```diff
@@ -76,10 +76,11 @@ export class HuiEntityEditor extends LitElement {
   protected _optionChanged(ev: Event): void {
     const target = ev.target! as EditorTarget;
```
Comment (balloob) [inline src/panels/lovelace/components/hui-entity-editor.ts]: Generally it's better to use `ev.currentTarget`, as that is the element you attached your handler to. `ev.target` could be a smaller element living inside the element you had your event listener attached to.

File: src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts [hunk 15]
```diff
@@ -0,0 +1,38 @@
+import { ConfigEntity } from "../../../cards/hui-glance-card";
+
+export class GlanceOptions {
+  private entityConf: ConfigEntity;
+
+  constructor(entityConf: ConfigEntity) {
+    const defaults = {
+      name: "",
+      icon: "",
+      tap_action: "more-info" as "more-info",
```
Comment (balloob) [inline src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts]: what to `as`?

File: src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+    const value = target.value || target.selected;
+    const index = target.parentElement!.index!;
+    const entity = this.entities![index];
+    const glanceOptions = new GlanceOptions(entity);
+
+    if (glanceOptions[target.configValue!] === value) {
+      return;
+    }
+
+    super._optionChanged(ev);
```
Comment (balloob) [inline src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: This seems like it might become a bit of a complicated flow of data. Why is this needed?

Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: The function in here checks the filter. and the super function will send the changed event

File: src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts (hunk 15 above)
Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts]: With out the `as` typescript doesnt like it. because we set tap_action in the interface to `"more-info" | "toggle" | "call-service"`

Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts]: https://github.com/Microsoft/TypeScript/issues/11465

Comment (balloob) [reply src/panels/lovelace/editor/config-elements/glance-card/glance-options.ts]: lame, but okay.

File: src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts (hunk 16 above)
Comment (balloob) [reply src/panels/lovelace/editor/config-elements/glance-card/hui-glance-entity-editor.ts]: I think that we can't know that the super event handler can always handle what we have rendered in options, we shouldn't code ourselves into a corner just yet. Only think about creating an abstraction for a pattern after having done 3 examples without an abstraction but with copy paste. Only then can we start to see if there is a pattern.

I would suggest that instead of `super._optionChanged(ev)`, we make something like `this._commitChangedOptions({[target.configValue]: value})`

**Comment by thomasloven** [comment]
Re: suggestions for how to handle service data.
A text input is good enough for /dev-service...
... but then again... that does have a "dev-" prefix...
